[JSC] VectorSwizzle should be optimized to DupElement when possible

Constellation opened the pull request:
#### aa2ac2e6fb27e0f59c36d9affbe0d2b2612f2fd5
<pre>
[JSC] VectorSwizzle should be optimized to DupElement when possible
<a href="https://bugs.webkit.org/show_bug.cgi?id=251241">https://bugs.webkit.org/show_bug.cgi?id=251241</a>
rdar://104722782

Reviewed by Justin Michaud.

This patch adds optimization for SIMD VectorSwizzle on ARM64.

1. If pattern only includes indexes for one side, we can convert 3-children VectorSwizzle to 2-children VectorSwizzle.
   This is better since we can freely use tbl v1 without register number constraints. And we can purge use of one side,
   potentially this can be removed if nobody is using it.
2. If pattern matches against DupElement, we convert it to VectorDupElement on ARM64. This can remove Const128 load for
   pattern, and use cheaper operation (dup element) instead of VectorSwizzle (tbl).

* JSTests/wasm/stress/simd-shuffle.js: Added.
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/assembler/ARM64Assembler.h:
(JSC::ARM64Assembler::dupElement):
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::vectorDupElement):
* Source/JavaScriptCore/assembler/MacroAssemblerRISCV64.h:
* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::vectorDupElement):
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3Opcode.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/b3/B3Opcode.h:
* Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
* Source/JavaScriptCore/b3/B3SIMDValue.h:
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::effects const):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueInlines.h:
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/jit/SIMDShuffle.h: Added.
(JSC::SIMDShuffle::isOnlyOneSideMask):
(JSC::SIMDShuffle::isI8x16DupElement):
(JSC::SIMDShuffle::isI16x8DupElement):
(JSC::SIMDShuffle::isI32x4DupElement):
(JSC::SIMDShuffle::isI64x2DupElement):
(JSC::SIMDShuffle::isI16x8Shuffle):
(JSC::SIMDShuffle::isI32x4Shuffle):
(JSC::SIMDShuffle::isI64x2Shuffle):
(JSC::SIMDShuffle::isIdentity):
(JSC::SIMDShuffle::isLargerElementShuffle):

Canonical link: <a href="https://commits.webkit.org/259464@main">https://commits.webkit.org/259464@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0089a40e78d91a636973a86bbc2f2b670d6a59f2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104977 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14057 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37870 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114242 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/174427 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15191 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4981 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97299 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/113259 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110737 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/11738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/94741 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/39255 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/93609 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/26366 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/80920 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/94826 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7396 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/27725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/92864 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/5128 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7491 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/4311 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/30265 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13547 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/47277 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/101559 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9279 "Built successfully") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/25336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3474 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -2438,6 +2438,14 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
 
     DEFINE_SIGNED_SIMD_FUNCS(vectorExtractLane);
 
+    void vectorDupElement(SIMDLane simdLane, TrustedImm32 lane, FPRegisterID src, FPRegisterID dest)
```
Is this needed if it is arm only?

Constellation replied:
To make it compiled, this is necessary.

justinmichaud reviewed (approved):
r=me

Constellation committed:
[JSC] VectorSwizzle should be optimized to DupElement when possible
https://bugs.webkit.org/show_bug.cgi?id=251241
rdar://104722782

Reviewed by Justin Michaud.

This patch adds optimization for SIMD VectorSwizzle on ARM64.

1. If pattern only includes indexes for one side, we can convert 3-children VectorSwizzle to 2-children VectorSwizzle.
   This is better since we can freely use tbl v1 without register number constraints. And we can purge use of one side,
   potentially this can be removed if nobody is using it.
2. If pattern matches against DupElement, we convert it to VectorDupElement on ARM64. This can remove Const128 load for
   pattern, and use cheaper operation (dup element) instead of VectorSwizzle (tbl).

* JSTests/wasm/stress/simd-shuffle.js: Added.
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/assembler/ARM64Assembler.h:
(JSC::ARM64Assembler::dupElement):
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::vectorDupElement):
* Source/JavaScriptCore/assembler/MacroAssemblerRISCV64.h:
* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::vectorDupElement):
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3Opcode.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/b3/B3Opcode.h:
* Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
* Source/JavaScriptCore/b3/B3SIMDValue.h:
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::effects const):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueInlines.h:
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/jit/SIMDShuffle.h: Added.
(JSC::SIMDShuffle::isOnlyOneSideMask):
(JSC::SIMDShuffle::isI8x16DupElement):
(JSC::SIMDShuffle::isI16x8DupElement):
(JSC::SIMDShuffle::isI32x4DupElement):
(JSC::SIMDShuffle::isI64x2DupElement):
(JSC::SIMDShuffle::isI16x8Shuffle):
(JSC::SIMDShuffle::isI32x4Shuffle):
(JSC::SIMDShuffle::isI64x2Shuffle):
(JSC::SIMDShuffle::isIdentity):
(JSC::SIMDShuffle::isLargerElementShuffle):

Canonical link: https://commits.webkit.org/259464@main
